### PR TITLE
[#50][feat] Add regexti instruction support

### PR DESCRIPTION
**Pull request by zhoujingya**
Initial commit

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 1]
```diff
@@ -96,10 +98,96 @@ bool RISCVExpandPseudo::expandMI(MachineBasicBlock &MBB,
...
+      break;
+    case RISCV::PseudoVMSNE_VI_IMM11:
+      MCID = &TII->get(RISCV::VMSNE_VI);
+      break;
+    case RISCV::PseudoVMSEQ_VI_IMM11:
+      MCID = &TII->get(RISCV::VMSEQ_VI);
+      break;
+  }
+
+  assert(MCID != nullptr && "Unexpected opcode");
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:140]: assert(MCID && "Unexpected opcode");

File: llvm/test/CodeGen/RISCV/VentusGPGPU/regexti.ll [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+define dso_local i1 @regexti12(i32 noundef %a) {
+entry:
+  ; CHECK-LABEL: regexti12:
+  ; CHECK: regexti	zero, zero, 1408
+  ; CHECK-NEXT: vmsne.vi	v0, v0, 15
+  %res = icmp ne i32 %a, -399
+  ret i1 %res
+}
```
Comment (zhoujingya) [inline llvm/test/CodeGen/RISCV/VentusGPGPU/regexti.ll:111]: Add test with regext with register numbers larger than 32

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 3]
```diff
@@ -96,10 +98,96 @@ bool RISCVExpandPseudo::expandMI(MachineBasicBlock &MBB,
   case RISCV::PseudoVMSGE_VI:
   case RISCV::PseudoVMSGEU_VI:
     return expandCompareSelect(MBB, MBBI, NextMBBI);
+  case RISCV::PseudoVOR_VI_IMM11:
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:101]: This way works too
it is better to set flags in instrtuction, you can refer to `vlw12.v` instruction, I remember I set flags before
Ih this way, it is more friendly when later we need to support other instruction

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 4]
```diff
@@ -96,10 +98,96 @@ bool RISCVExpandPseudo::expandMI(MachineBasicBlock &MBB,
   case RISCV::PseudoVMSGE_VI:
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:98]: what above means is that: "MBBI.isVOPIMM11()" such form

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 5]
```diff
@@ -96,10 +98,96 @@ bool RISCVExpandPseudo::expandMI(MachineBasicBlock &MBB,
...
+  MBBI->setDesc(*MCID);
+
+  int64_t Imm = 0;
+  signed LowImm = 0;
+  signed Offsets = 0;
+
+  for (unsigned i = 0; i < MBBI->getNumOperands(); ++i) {
+    MachineOperand &MO = MBBI->getOperand(i);
+
+    // deal with 4 ext reg.
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:150]: More friendly comment is needed

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 6]
```diff
@@ -96,10 +98,96 @@ bool RISCVExpandPseudo::expandMI(MachineBasicBlock &MBB,
...
+      uint16_t RegEncodingValue = TRI->getEncodingValue(MO.getReg());
+      if (RegEncodingValue > 31) {
+        int Pos = MBBI->getDesc().getOperandConstraint(i, MCOI::CUSTOM);
+        assert(Pos != -1 && "Out of range[0, 31] register operand custom "
+                            "constraint that must be present.");
+        Offsets |= (RegEncodingValue >> 5 & 0x7) << (3 * Pos);
+      }
+    }
+
+    if (MO.isImm()) {
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:162]: Format?

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 7]
```diff
@@ -96,10 +98,96 @@ bool RISCVExpandPseudo::expandMI(MachineBasicBlock &MBB,
...
+        Imm &= 0b01111111111;
+        Imm |= 0b10000000000;
+      }
+      
+      LowImm   = (Imm & 0b00000011111);
+      Offsets |= (Imm & 0b11111100000);
+
+      MO.ChangeToImmediate(LowImm);
+    }
+    // how to guarantee that the statement execute only once.
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:179]: What does this comment mean

File: llvm/lib/Target/RISCV/VentusInstrFormatsV.td [hunk 8]
```diff
@@ -62,6 +62,27 @@ class RVInstSetVLi<dag outs, dag ins, string opcodestr, string argstr>
...
+}
+
+class PseudoVOPIVIMM11<SDPatternOperator Op> :
+        Pseudo<(outs VGPR:$vd), (ins VGPR:$vs2, simm11:$imm11),
+        [(set (XLenVT VGPR:$vd), (Op simm11:$imm11, (XLenVT VGPR:$vs2)))]> {
+  // let usesCustomInserter = 1;
+  let hasSideEffects = 1;
+  let mayLoad = 0;
+  let mayStore = 0;
+  let IsVOPIMM11 = 1;
```
Comment (zhoujingya) [inline llvm/lib/Target/RISCV/VentusInstrFormatsV.td:83]: let hasSideEffects = 1, mayLoad = 0, mayStore = 0, IsVOPIMM11 = 1 in {
class PseudoVOPVIIMM11<SDPatternOperator Op> :
        Pseudo<(outs VGPR:$vd), (ins VGPR:$vs2, simm11:$imm11),
        [(set (XLenVT VGPR:$vd), (Op (XLenVT VGPR:$vs2), simm11:$imm11))]>;
class PseudoVOPIVIMM11<SDPatternOperator Op> :
        Pseudo<(outs VGPR:$vd), (ins VGPR:$vs2, simm11:$imm11),
        [(set (XLenVT VGPR:$vd), (Op simm11:$imm11, (XLenVT VGPR:$vs2)))]>

}

**Review by wangqinfan — APPROVED** [review]
LGTM

**Pull request merged**